### PR TITLE
Enable observe training entropy without computing entropy loss

### PR DESCRIPTION
**Pull request by zyzshishui**
## Motivation
Algorithm bros want to observe entropy from the training forward pass even when the entropy is not included in calculating loss. The previous `--entropy-coef 0.0` path skipped training entropy entirely, leaving only rollout-time entropy available for dashboards.

## Usage
```
--entropy-coef 0.0 --observe-training-entropy
```
This logs entropy_loss from the training logits as a detached metric. It does not add the entropy to the loss and does not attach the observed entropy to backward.

### CODE REVIEW

**Review by guapisolo — APPROVED** [review]
LGTM. need code owner review.

**Comment by guapisolo** [comment]
CC said the entropy calc definitely introduce some overhead.. so it should be never calculated under `--entropy-coef 0.00`, and can be re-opened by the new `observe-training-entropy` flag.

@zyzshishui Do you think algorithm people happy to workaround with `--entropy-coef 1e-15` (extremely tiny won't degrade training, but can still observe the entropy metrics?)

**Comment by zyzshishui** [comment]
> CC said the entropy calc definitely introduce some overhead.. so it should be never calculated under `--entropy-coef 0.00`, and can be re-opened by the new `observe-training-entropy` flag.
> 
> @zyzshishui Do you think algorithm people happy to workaround with `--entropy-coef 1e-15` (extremely tiny won't degrade training, but can still observe the entropy metrics?)

I dont think they will be happy...

**Comment by guapisolo** [comment]
> > CC said the entropy calc definitely introduce some overhead.. so it should be never calculated under `--entropy-coef 0.00`, and can be re-opened by the new `observe-training-entropy` flag.
> > @zyzshishui Do you think algorithm people happy to workaround with `--entropy-coef 1e-15` (extremely tiny won't degrade training, but can still observe the entropy metrics?)
> 
> I dont think they will be happy...

ok fine.

File: miles/backends/training_utils/loss_hub/math_utils.py [hunk 1]
```diff
@@ -941,7 +953,13 @@ def _calculate_log_probs_and_entropy_true_on_policy(
 
     entropy = None
     if with_entropy:
-        probs = log_probs_full.exp()
-        entropy = -(probs * log_probs_full).sum(dim=-1)
+        entropy_log_probs = log_probs_full if entropy_requires_grad else log_probs_full.detach()
```
Comment (yueming-yuan) [inline miles/backends/training_utils/loss_hub/math_utils.py:956]: qq: curious why don't we directly replace `with_entropy` with `entropy_requires_grad`? like
```
if entropy_requires_grad:
        probs = log_probs_full.exp()
        entropy = -(probs * log_probs_full).sum(dim=-1)
```

Comment (zyzshishui) [reply miles/backends/training_utils/loss_hub/math_utils.py:956]: `with_entropy` decides whether entropy is computed, and `entropy_requires_grad` decides whether that entropy participates in autograd. Our algo bro wants observe-only case ——`with_entropy=True` and `entropy_requires_grad=False`

Comment (zyzshishui) [reply miles/backends/training_utils/loss_hub/math_utils.py:956]: Adding entropy into computing graph is memory consuming, and leads to OOM which could be avoided by this patch

Comment (yueming-yuan) [reply miles/backends/training_utils/loss_hub/math_utils.py:956]: oh sorry i was sb...
just one more question: it seems if we already detach the `entropy_log_probs`, we don't need the  `with torch.no_grad():` branch? 
in the way below i think it's also correct (and shorter)
```
if with_entropy:
        entropy_log_probs = log_probs_full if entropy_requires_grad else log_probs_full.detach()
        probs = entropy_log_probs.exp()
        entropy = -(probs * entropy_log_probs).sum(dim=-1)
```

Comment (zyzshishui) [reply miles/backends/training_utils/loss_hub/math_utils.py:956]: I was sb... fixed

**Review by yueming-yuan — APPROVED** [review]
